Some tweaks and additions to Video/Photo capture.

dogmaphobic opened the pull request:
- Establishes that a `CAMERA_IMAGE_CAPTURED` message is sent after each image captured as a result of `MAV_CMD_IMAGE_START_CAPTURE`.
- Clarifies that `MAV_CMD_VIDEO_START_CAPTURE` and `MAV_CMD_VIDEO_STOP_CAPTURE ` means “Recording”.
- Add the ability to request CAMERA_CAPTURE_STATUS messages to be sent automatically during video recording.
- Add a field to `CAMERA_CAPTURE_STATUS` telling the current recording time.
- Add a field to `CAMERA_CAPTURE_STATUS` telling the current remaining storage capacity.
- Changed the `file_path` in `CAMERA_IMAGE_CAPTURED` to `file_url` as this is more appropriate. A local storage path is of limited use. But if the camera supports image requests through an HTTP interface, this can be used to send the URL to obtain the image.

WickedShell commented:
I realize I'm late to this series of messages but I had a couple of questions primarily about CAMERA_IMAGE_CAPTURED. ArduPilot has the CAMERA_FEEDBACK message as well which looks like it's purpose overlaps with the CAMERA_IMAGE_CAPTURED. Do you envision always emitting a CAMERA_IMAGE_CAPTURED if the aircraft automatically triggered a non mavlink capable camera? If so there are a couple of questions I have about it, or what the replacement for CAMERA_FEEDBACK would be in that scenario. The most important one that is missing is the index for counting the number of photos attempted/succeeded. (Which the ArduPilot message only has one of, and needs to be extended for)

The second is that altitude AMSL is actually a poor choice rather then WGS84. Your lat/lng alt really needs to be in the same reference frame or it complicates handling the data, and while you may want to use a different end reference frame at least WGS84 provides you with a reasonable standard that everything should be able to be referenced from. (ArduPilot has a doc issue in that we say AMSL on the CAMERA_FEEDBACK message but we only provide WGS84 instead).

dogmaphobic commented:
I don't have an opinion regarding the altitude. I would let those who have a stronger preference for one or the other to chime in.

I agree that an image count (image index n out of x) would be nice to have but I'm not sure how much data you can fit in these messages. Ideally, it would have a total count, current index, and a success or failure boolean for the current index.

>Do you envision always emitting a CAMERA_IMAGE_CAPTURED if the aircraft automatically triggered a non mavlink capable camera? 

I don't see why not. There is also the case of photos taken as a result of mission commands such as `MAV_CMD_DO_SET_CAM_TRIGG_DIST` for instance. The trick is how to document these as once we establish the `CAMERA_IMAGE_CAPTURED` message will be sent as a result of *any* still captured, it should be made clear everywhere.

dogmaphobic commented:
>ArduPilot has a doc issue in that we say AMSL on the CAMERA_FEEDBACK message but we only provide WGS84 instead

That's what breaks a protocol and makes it useless. Sort of having an HTML standard and Microsoft's own "peculiar" way of doing things differently and breaking everything in the process.

WickedShell commented:
We need to fix the comment on it on the ArduPilot document. I'd guess the original message author hadn't considered what it was meant to hold, but any absolute altitude from ArduPilot is expected to be WGS84 regardless of any XML comment to the contrary.

dogmaphobic commented:
I've added an `image_index` field with the (zero based) index for the current image. That would be the count of images (-1) taken (successfully or otherwise) since started.
I've also added a boolean (`capture_result`) indicating the success or failure capturing the image.
I figured the total count expected would be better handled at the GCS end. In other words, this index is x out of n where n is defined or easily computed by the GCS.

WickedShell commented:
The GCS can't be expected to know how many were attempted, as the aircraft could be creating the attempt commands and the result packets might be lost on the way to the GCS. To really cover that you need an index as well as a count for the total number attempted, otherwise packet loss will defeat the attempted account.

dogmaphobic commented:
> The GCS can't be expected to know how many were attempted

I guess we're talking about two different things, which brings up a new issue. The way I had in mind, you set the camera to take n number of pictures with an x interval, or "time lapse" (and/or distance in between when dealing with "mission"). There are no provisions for subsequent attempts. The image was either successfully taken or not. I'm thinking mostly about cameras with a direct, digital interface. The only case where a capture would fail would be in the case of full storage. The `image_index` field above is essentially a frame counter.

Are you talking about cameras with mechanical (e.g. PWM) switches? I don't think that's the intent of these particular set of messages, hence the `file_url` field. Mechanical switches are mostly obsolete at this point in time and no modern system would use it. These messages are meant for intelligent cameras that provide a digital interface for controlling them. You communicate either directly or through a (companion computer) controller.

With that said, there is no harm in adding the fields you feel are necessary. I would like however, to have more feedback on this regarding future technologies instead of having to deal with legacy, mostly hobby systems as those are far more flexible in how these things get implemented. By their own nature, hobby systems are mostly one-offs where "workaround" is the norm. These are standards for intelligent image/video capture devices.

@LorenzMeier @julianoes Your thoughts about this?

WickedShell commented:
Well thats why I was unsure if this was meant to be orthogonal to the CAMERA_FEEDBACK message APM provides. Even with whats currently called "relay" control from the autopilot the camera can fail to take a photo for a variety of reasons (particularly if it was simply not ready to store the next photo). Even with cameras with digital protocols this can happen. In that case the GCS really benefits from the information that some number of attempted photos had failed.

The per message telling you about failure works well for indicating if a command failed, but if you lose the packet there is no way to tell if the photo attempt was successful. Even with a camera that is itself MAVLink capable it could fail for a variety of reasons (interval to fast, failure to solve focus if it was in autofocus (not that you should be mapping this way, but it's not inconceivable) the camera was busy with an internal operation (IE FFC's).

dogmaphobic commented:
Got it. I'm looking at APM's `CAMERA_FEEDBACK_FLAGS`.

I see a *failure* value but nothing regarding attempts at retaking a failed attempt. Are you suggesting something new that would handle these cases or am I missing something when looking at what APM has already? Are you saying there should be a flag telling the camera to keep trying until it reaches the next trigger event?

For mapping, or other periodic photographic functions, when planning the flight, the GCS will give you an estimate of time in between shots. That should give you some indication on the probability of errors if the interval is too short. I would imagine that's when you want to make sure your margins are within the camera's capabilities. The GCS will also give you a precise number of shots that will be taken. That's what I meant above about not needing a "total_count" field within the message.

With that said, I'm open to anything. Especially if it's something common as opposed to something that satisfies corner cases only.

My background is cinematography and/or planned "single" stills. I'm certainly no expert in what you are asking about so I can't claim I know exactly what is needed.

WickedShell commented:
Ah sorry I should have been clearer. The lack of the count is a weakness of the CAMERA_FEEDBACK message APM currently uses. It's been on my to fix list by extending it with MAVLink 2 (I need to get to the APM code side first however).

So the specific problem that I'm looking to cover here is related to doing photogrammetry of large areas. (~2+ hours of mapping, with shutter synchronization with the camera). There are a variety of reasons that you can not receive a shutter signal when you attempt to take a photo. They cover the envelope mentioned before. There are 3 specific problems that I'd like to see covered, that I've had problems with in the past mapping large areas.

1) We have a intermittent (or extremely low bandwidth) link to the aircraft that can't send every message to the GCS, so there is no positive confirmation on every photo attempt being successful. The total would allow us to detect this situation and inform the user, and inform them in a more meaningful manner then just a health flag. The other problem with this is even if we send a failure flag that doesn't provide the GCS with a great indication of what our failure rate is, however a count allows us to assess the failure rate. 
2) We never received the synchronization signal. The problem with some setups is the camera might take awhile to take a photo. It's hard to know when to send a failure message to the GCS if the delay on responding to the event is variable or potentially long. I'm just not sure when you would expect the autopilot to report a failure in this scenario.
3) While you are correct that the GCS can predict the number of shots taken it can't be certain if it will fail to keep up. Valid reasons for this are wind changes, momentary downwind turns that are within the cameras burst ability, and the compressibility of the scene (if compression is enabled) has a very large impact on the buffer size. Depending on the mission you can accept a low failure rate in certain areas, and with the lossy nature of the GCS to the aircraft this is the best way to ensure that we are communicating a level of failure to the GCS.

Does that make sense? I might be explaining this poorly.

dogmaphobic commented:
Alright, what you describe is both extremely specialized and complex. In my experience, trying to generalize something like this along with a regular "taking photos" task is prone to make the former not good enough and the latter way more complex than it needs to be.

What you need is a very specific set of messages that address the need of photogrammetry within the context of unreliable camera performance (missed captures for a variety of reasons) and unreliable message delivery (long distances).

My first reaction is that ought to be handled at the vehicle side as opposed to on the GCS side. A software running on the vehicle is far more capable of handling these situations than depending on a roundtrip communication with the GCS so the GCS can figure out what to do. On such "missions", the role of the GCS should be to monitor the progress. The intelligence should be within the vehicle. As such, the proposed set of messages are mostly sufficient. It is up to the computer running on the vehicle to handle all these exceptions and provide updates to the GCS using simpler messages.

For a serious photogrammetry task, I would imagine it would be far more cost effective to develop a system where the intelligence is embedded within the vehicle. The vehicle (rather, the companion computer connected to the flight controller--the "A" of UAV) would handle these tasks and communicate with the GCS through MAVLink using the messages described here. I don't think it makes a whole lot of sense to define a "common" set of messages to handle such extremely specialized tasks.

Given that this is a specialized usage, it might make more sense to develop a new set of messages to supplement the core "common" set. One that interacts with a component within the vehicle handling these tasks as opposed to something you communicate directly with the flight controller.

With all that said, this is just an opinion from someone who honestly, has no business dictating any of this :P

julianoes commented on message_definitions/v1.0/common.xml:
```diff
@@ -3850,7 +3853,9 @@
-      <field type="char[210]" name="file_path">File path of image taken.</field>
+      <field type="int32_t" name="image_index">Zero based index of this image (count since started -1)</field>
+      <field type="int8_t" name="capture_result">Boolean indicating success (1) or failure (0) while capturing this image.</field>
+      <field type="char[210]" name="file_url">URL of image taken. Either local storage or http://foo.jpg if camera provides an HTTP interface.</field>
```
This gets us to 248 bytes if I counted correctly. Is that on purpose?

dogmaphobic replied:
Nope, I'm not sure what the limit is. The 210-byte allocation for the file path was already there, which I found strange. I simply added 5 bytes indiscriminately assuming either CI or you would tell me it went past the limit :P

julianoes replied:
The limit is 255 bytes, so that's still fine.

julianoes commented on message_definitions/v1.0/common.xml:
```diff
@@ -3850,7 +3853,9 @@
       <field type="int32_t" name="relative_alt">Altitude above ground in meters, expressed as * 1E3 where image was taken</field>
       <field type="float[4]" name="q">Quaternion of camera orientation (w, x, y, z order, zero-rotation is 0, 0, 0, 0)</field>
-      <field type="char[210]" name="file_path">File path of image taken.</field>
+      <field type="int32_t" name="image_index">Zero based index of this image (count since started -1)</field>
```
Since a flight started or since a recording started?

dogmaphobic replied:
Good point. Since capture started.

dogmaphobic replied:
Come to think of it, it depends. We need to think about this. If this message is sent only when images are captured in response to a `MAV_CMD_IMAGE_START_CAPTURE` message, then it would make sense to for the count to be reset when it receives a `MAV_CMD_IMAGE_START_CAPTURE` message. However, if the `CAMERA_IMAGE_CAPTURED` is to be sent after any image is captured by whatever means, such as a result of `MAV_CMD_DO_SET_CAM_TRIGG_DIST`, in that case, it makes no sense. In this case, it would be better for it to be a global count, which is reset when the vehicle is armed.

The real question: should `CAMERA_IMAGE_CAPTURED` be sent every time an image is captured or only during a `MAV_CMD_IMAGE_START_CAPTURE` session? My instinct says yes and the counter reset when the vehicle is armed.

julianoes replied:
> My instinct says yes (sent for all captures) and the counter reset when the vehicle is armed.

I would agree with this behaviour.

julianoes commented on message_definitions/v1.0/common.xml:
```diff
@@ -3839,6 +3840,8 @@
       <field type="uint16_t" name="image_resolution_v">Image resolution in pixels vertical</field>
       <field type="uint16_t" name="video_resolution_h">Video resolution in pixels horizontal</field>
       <field type="uint16_t" name="video_resolution_v">Video resolution in pixels vertical</field>
+      <field type="uint32_t" name="recording_time_s">Time in seconds since recording started</field>
```
What about recording time in ms?

dogmaphobic replied:
Why not indeed.

dogmaphobic replied:
Done. My original thought was twofold: Most cameras won't give you a better than 1 second resolution and it's unlikely you would have these messages sent out in high enough frequency to justify millisecond accuracy. I did think about sending seconds using a float but that would half the amount of seconds we could store (albeit far beyond what a regular camera could record in one go).

LorenzMeier commented:
@WickedShell Its worth re-checking your assumption about WGS84 altitudes, because by default GPS receivers output WGS84 position and *AMSL* altitude. It would be in fact surprising if APM would indeed be using WGS84 and the MAVLink spec specifically calls for *AMSL*.
